test/compile-error-test/CMakeLists.txt

```diff
@@ -1,77 +1,128 @@
-set(CMAKE_REQUIRED_FLAGS ${CXX_STANDARD_FLAG} ${PEDANTIC_COMPILE_FLAGS})
-
-function (generate_source result fragment)
-  set(${result} "
-  #define FMT_HEADER_ONLY 1
-  #include \"fmt/format.h\"
-  int main() {
-    ${fragment}
-  }
-  " PARENT_SCOPE)
+function (expect_compile name code_fragment)
+  cmake_parse_arguments(EXPECT_COMPILE "ERROR;FORMAT;XCHAR" "" "" ${ARGN})
+  set(test_names_local_copy "${test_names}")
+  list(APPEND test_names_local_copy "${name}")
+  set(test_names "${test_names_local_copy}" PARENT_SCOPE)
+  set(header "fmt/core.h")
+  if (EXPECT_COMPILE_FORMAT)
+    set(header "fmt/format.h")
+  elseif(EXPECT_COMPILE_XCHAR)
+    set(header "fmt/xchar.h")
+  endif()
+  configure_file("${CMAKE_CURRENT_SOURCE_DIR}/source.cpp.in" "${CMAKE_CURRENT_BINARY_DIR}/tests/${name}.cpp")
+  if (EXPECT_COMPILE_ERROR)
+    set(tests_${name}_error_expected 1 PARENT_SCOPE)
+  else()
+    set(tests_${name}_error_expected 0 PARENT_SCOPE)
+  endif()
 endfunction ()
 
-function (expect_compile code)
-  generate_source(source "${code}")
-  check_cxx_source_compiles("${source}" compiles)
-  if (NOT compiles)
-    set(error_msg "Compile error for: ${code}")
-  endif ()
-  # Unset the CMake cache variable compiles. Otherwise the compile test will
-  # just use cached information next time it runs.
-  unset(compiles CACHE)
-  if (error_msg)
-    message(FATAL_ERROR ${error_msg})
-  endif ()
-endfunction ()
+function (run_tests)
+  set(cmake_lists_file_projects "")
+  foreach(test_name IN LISTS test_names)
+    string(CONCAT cmake_lists_file_projects "${cmake_lists_file_projects}" "
+      add_executable(test-${test_name} ${test_name}.cpp)
+      target_link_libraries(test-${test_name} PRIVATE fmt::fmt)
+    ")
+  endforeach()
+  configure_file("${CMAKE_CURRENT_SOURCE_DIR}/CMakeLists.txt.in" "${CMAKE_CURRENT_BINARY_DIR}/tests/CMakeLists.txt")
+
+  set(build_directory "${CMAKE_CURRENT_BINARY_DIR}/tests/build")
+  file(REMOVE_RECURSE "${build_directory}")
+  file(MAKE_DIRECTORY "${build_directory}")
+  execute_process(
+    COMMAND
+      "${CMAKE_COMMAND}"
+      "-DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}"
+      "-DCMAKE_CXX_FLAGS=${CMAKE_CXX_FLAGS}"
+      "-DCMAKE_CXX_STANDARD=${CMAKE_CXX_STANDARD}"
+      "-DCMAKE_GENERATOR=${CMAKE_GENERATOR}"
+      "-DCMAKE_MAKE_PROGRAM=${CMAKE_MAKE_PROGRAM}"
+      "-DFMT_DIR=${FMT_DIR}"
+      "${CMAKE_CURRENT_BINARY_DIR}/tests"
+    WORKING_DIRECTORY "${build_directory}"
+    RESULT_VARIABLE resultVar
+    OUTPUT_VARIABLE outputVar
+    ERROR_VARIABLE errorVar)
+  if (NOT resultVar EQUAL 0)
+    message(FATAL_ERROR "Unable to configure:\n${errorVar}")
+  endif()
+
+  execute_process(
+    COMMAND
+      "${CMAKE_COMMAND}" --build "${build_directory}" --target fmt
+    WORKING_DIRECTORY "${build_directory}"
+    RESULT_VARIABLE resultVar
+    OUTPUT_VARIABLE outputVar
+    ERROR_VARIABLE errorVar)
+  if (NOT resultVar EQUAL 0)
+    message(FATAL_ERROR "Unable to build {fmt} library:\n${errorVar}")
+  endif()
```

Comment:
Is this needed? Won't fmt be built as a dependency below?

Reply:
It does build by dependency, but I decided it would be nice to have the library's own dedicated output in case of error during the library build. Also, there is a mostly imaginary problem (which I want to write down anyway): in case of errors during the library build, all `expect_compile(... ERROR)` calls would pass unconditionally, while `expect_compile(...)` would fail. Considering everything above, I'm not sure it's needed as long as we have those checks:
https://github.com/fmtlib/fmt/blob/c28753cadd1f8606215f8311309dacd045a24225/test/compile-error-test/CMakeLists.txt#L86-L88

Reply:
Since the error in building the library will be caught, I think this is not needed so let's remove to keep the config simpler.

Reply:
Removed